src/Voron/Impl/Journal/WriteAheadJournal.cs

```diff
@@ -946,85 +887,119 @@ private void CallPagerSync()
-                        _fsyncLockTaken = _parent._fsyncLock.Wait(0);
-                        if (_fsyncLockTaken == false)
-                        {
-                            // probably another sync taking place right now, let us schedule another one, just in case
-                            _parent._waj._env.QueueForSyncDataFile();
-                            return false;
-                        }
-                        _currentTotalWrittenBytes = Interlocked.Read(ref _parent._totalWrittenButUnsyncedBytes);
-                        _lastSyncedJournal = _parent._lastFlushedJournalId;
-                        _lastSyncedTransactionId = _parent._lastFlushedTransactionId;
-                        _parent._lastFlushedJournal.SetLastReadTxHeader(_lastSyncedTransactionId, ref _transactionHeader);
-                        if (_lastSyncedTransactionId != _transactionHeader.TransactionId)
-                        {
-                            VoronUnrecoverableErrorException.Raise(_parent._waj._env,
-                                $"Error syncing the data file. The last sync tx is {_lastSyncedTransactionId}, but the journal's last tx id is {_transactionHeader.TransactionId}, possible file corruption?"
-                            );
-                        }
+                    foreach (var kvp in _journalsToDelete)
+                    {
+                        _parent._journalsToDelete.Remove(kvp.Key);
+                    }
+                }
 
-                        _parent._lastFlushedJournal = null;
+                public void Dispose()
+                {
+                    if (_fsyncLockTaken)
+                        _parent._fsyncLock.Release();
+                }
+            }
 
-                        foreach (var toDelete in _parent._journalsToDelete)
-                        {
-                            if (toDelete.Key > _lastSyncedJournal)
-                                continue;
+            private class LockTaskResponsible
+            {
+                private readonly object _lock;
+                private readonly CancellationToken _token;
+                private Action _task;
 
-                            _journalsToDelete.Add(toDelete);
-                        }
+                private readonly ManualResetEventSlim _waitForTaskToBeDone = new ManualResetEventSlim();
 
-                        _parent._waj._env.Options.SetLastReusedJournalCountOnSync(_journalsToDelete.Count);
+                public LockTaskResponsible(object @lock, CancellationToken token)
+                {
+                    _lock = @lock;
+                    _token = token;
+                }
 
-                        foreach (var kvp in _journalsToDelete)
+                public bool WaitForTaskToBeDone(Action task)
+                {
+                    var isLockTaken = false;
+                    try
+                    {
+                        _waitForTaskToBeDone.Reset();
+                        _task += task;
```

Comment:
We do `+=` but never `-=` and we have just one instance of `LockTaskResponsible` so we only append tasks and always execute all of them.

Reply:
In `RunTaskIfNotAlreadyRan` method the `_task` field assigned to `null` [<sup>ref</sup>](RunTaskIfNotAlreadyRan)